### ISSUE TITLE
Speedup for drago

### ISSUE THREAD

**Issue by heckflosse** [LuminanceHDR/LuminanceHDR#80]
The intention of this issue is to make a step by step speedup for drago tonemapper.
To follow the optimizations you can clone my [fork of LuminanceHDR](https://github.com/heckflosse/LuminanceHDR) and build the drago_speedup branch.
I will include one commit per post and describe the optimizations.
Finally I will make a pr.
All measurements will use the same 24Mpixel file

Let's start with the first commit 65cc206,  which just adds some timing code: 
Now we have an initial measure for drago which is at my machine:

~ 11.8 seconds

**Comment by heckflosse** [comment]
There's a loop which runs a bit cache-unfriendly. Changed in fcb3df4

~ 11.2 seconds

**Comment by heckflosse** [comment]
Used openmp for the loop from last post in 61e8163

~ 11 seconds

**Comment by heckflosse** [comment]
Use xlogf() from sleed instead of log() in 812b8bf

~ 7.6 seconds

**Comment by heckflosse** [comment]
Parallelized and vectorized one function in ede4c39

~ 6.7 seconds

**Comment by heckflosse** [comment]
Another use case of sleef functions in 8eb24db

~ 2.4 seconds

**Comment by heckflosse** [comment]
Next one with 136b157

~ 0.5 seconds

**Comment by heckflosse** [comment]
Another final speedup will follow tomorrow

**Comment by Benitoite** [comment]
https://filebin.net/m0ol4i4hhlzlm1gm/Luminance_HDR_2.5.2-SNAPSHOT-dragotest1.dmg
Nice job Ingo, `drago_speedup` working fine on mac.

**Comment by fcomida** [comment]
Good job, I have some more speed up vectorizing drago03 main loop. Can we do the same in the final loop in pfstmo_drago03.cpp? There is a "if (yr != 0.f)" ...

**Comment by heckflosse** [comment]
@fcomida That's possible:

either
```
            vfloat yrv = LVFU(Yr(x, y));
            vfloat scalev = vself(vmaskf_neq(yrv, ZEROV), LVFU(L(x,y)) / yrv, ZEROV);
```

or even faster
```
            vfloat yrv = LVFU(Yr(x, y));
            vfloat scalev = vselfzero(vmaskf_eq(yrv, ZEROV), LVFU(L(x,y)) / yrv);
```

untested

**Comment by fcomida** [comment]
That was your change for today. Post your version as well for comparison (for me to learn ;) )

**Comment by heckflosse** [comment]
Ok, I just found a bug in my changes. I will fix it and push in about an hour or so.

**Comment by heckflosse** [comment]
Final speedups and a bugfix with ccf309e 

< 0.3 seconds

**Comment by heckflosse** [comment]
I just pushed a bugfix...

**Comment by fcomida** [comment]
Working on durand, from 30sec to 16.5sec and I notice in drago you can mix
float and vfloat and it's faster (biasPv, dividerv ....).

**Comment by heckflosse** [comment]
@fcomida Franco, in gcc it is possible to write stuff like
```
vfloat x = y + 2.f;
```

instead of
```
vfloat twov = F2V(2.f);
vfloat x = y + twov;
```

but at least in past that did not work in clang...

**Comment by fcomida** [comment]
yeah and it is slightly faster, I'll try compiling with clang...

**Comment by heckflosse** [comment]
I played a bit with mantiuk06 and got already double speed. But it's not ready to make a branch atm

**Comment by fcomida** [comment]
You are right, it doesn't compile with clang.

**Comment by fcomida** [comment]
I have clang version 4.0.1

**Comment by heckflosse** [comment]
I don't have clang at all ;-)

**Comment by Benitoite** [comment]
I have clang 3.9.1... :)

**Comment by heckflosse** [comment]
@fcomida Franco, about durand: Did you have a look at fastbilateral.cpp? There are some nice speedups possible. I can post a patch if you like.

**Comment by fcomida** [comment]
I fixed clang compilation, right now I have ~ 15x speed gain and space for
more improvement, I didn't get it quite right though, image is too bright
:( . Some stupid mistake that
I'll check tomorrow.
If you like post your implementation, no problem at all.

**Comment by heckflosse** [comment]
Which tmo is that ~ 15x speed gain?

**Comment by heckflosse** [comment]
@fcomida 

Here's my patch for `fastbilateral.cpp`

```patch
diff --git a/src/TonemappingOperators/durand02/fastbilateral.cpp b/src/TonemappingOperators/durand02/fastbilateral.cpp
index 726833c4..e3d40b88 100644
--- a/src/TonemappingOperators/durand02/fastbilateral.cpp
+++ b/src/TonemappingOperators/durand02/fastbilateral.cpp
@@ -43,6 +43,10 @@
 #define likely(x) (x)
 #define unlikely(x) (x)
 #endif
+#define BENCHMARK
+#include "../../StopWatch.h"
+#include "../../sleef.c"
+#include "../../opthelper.h"
 
 using namespace std;
 
@@ -76,6 +80,7 @@ class GaussianBlur {
     }
 
     void blur(const pfs::Array2Df &I, pfs::Array2Df &J) {
+        BENCHFUN
         int x, y;
 
         int nx = I.getCols();
@@ -85,18 +90,21 @@ class GaussianBlur {
         int ox = nx;
         int oy = ny / 2 + 1;  // saves half of the data
 
-        for (y = 0; y < ny; y++)
-            for (x = 0; x < nx; x++) source[x * ny + y] = I(x, y);
+        #pragma omp parallel for
+        for (x = 0; x < nx; x++)
+            for (y = 0; y < ny; y++)
+                source[x * ny + y] = I(x, y);
 
         fftwf_execute(fplan_fw);
 
         // filter
         float sig = nx / (2.0f * sigma);
         float sig2 = 2.0f * sig * sig;
+        #pragma omp parallel for
         for (x = 0; x < ox / 2; x++)
             for (y = 0; y < oy; y++) {
                 float d2 = x * x + y * y;
-                float kernel = exp(-d2 / sig2);
+                float kernel = xexpf(-d2 / sig2);
 
                 freq[x * oy + y][0] *= kernel;
                 freq[x * oy + y][1] *= kernel;
@@ -106,6 +114,7 @@ class GaussianBlur {
 
         fftwf_execute(fplan_in);
 
+        #pragma omp parallel for
         for (x = 0; x < nx; x++)
             for (y = 0; y < ny; y++) J(x, y) = source[x * ny + y] / nsize;
     }
@@ -243,17 +252,9 @@ void fastBilateralFilter(const pfs::Array2Df &I, pfs::Array2Df &J,
     }
 
     pfs::Array2Df *JJ;
-    //  if( downsample != 1 )
-    //    JJ = new pfs::Array2D(w,h);
-
-    //  w /= downsample;
-    //  h /= downsample;
 
     int sizeZ = w * h;
-    //  pfs::Array2D* Iz = new pfs::Array2D(w,h);
-    //  downsampleArray(I,Iz);
     const pfs::Array2Df *Iz = &I;
-    //  sigma_s /= downsample;
 
     pfs::Array2Df jJ(w, h);
     pfs::Array2Df jG(w, h);
@@ -272,11 +273,29 @@ void fastBilateralFilter(const pfs::Array2Df &I, pfs::Array2Df &J,
 
         float jI = minI + j * stepI;  // current intensity value
 
-        for (int i = 0; i < sizeZ; i++) {
-            float dI = (*Iz)(i)-jI;
-            jG(i) = exp(-(dI * dI) / (sigma_r * sigma_r));
-            jH(i) = jG(i) * I(i);
+#pragma omp parallel
+{
+#ifdef __SSE2__
+        vfloat sqrsigma_rv = F2V(sigma_r * sigma_r);
+        vfloat jIv = F2V(jI);
+#endif
+        #pragma omp for
+        for (int i = 0; i < h; i++) {
+            int j = 0;
+#ifdef __SSE2__
+            for (; j < w-3; j+=4) {
+                vfloat dIv = LVFU((*Iz)(j, i)) - jIv;
+                STVFU(jG(j, i), xexpf(-(dIv * dIv) / sqrsigma_rv));
+                STVFU(jH(j, i), LVFU(jG(j, i)) * LVFU(I(j, i)));
+            }
+#endif
+            for (; j < w; j++) {
+                float dI = (*Iz)(j, i) - jI;
+                jG(j, i) = xexpf(-(dI * dI) / (sigma_r * sigma_r));
+                jH(j, i) = jG(j, i) * I(j, i);
+            }
         }
+}
 
         gaussian_blur.blur(jG, jK);
         gaussian_blur.blur(jH, jH);
@@ -284,12 +303,13 @@ void fastBilateralFilter(const pfs::Array2Df &I, pfs::Array2Df &J,
         //    convolveArray(jG, sigma_s, jK);
         //    convolveArray(jH, sigma_s, jH);
 
-        for (int i = 0; i < sizeZ; i++)
-            if (likely(jK(i) != 0.0f))
-                jJ(i) = jH(i) / jK(i);
-            else
-                jJ(i) = 0.0f;
-
+        #pragma omp parallel for
+        for (int i = 0; i < h; i++) {
+            for (int j = 0; j < w; j++) {
+                float temp = jK(j, i);
+                jJ(j, i) = temp != 0.f ? jH(j, i) / temp : 0.f;
+            }
+        }
         //  if( downsample == 1 )
         JJ = &jJ;  // No upsampling is necessary
                    //    else
@@ -297,6 +317,7 @@ void fastBilateralFilter(const pfs::Array2Df &I, pfs::Array2Df &J,
 
         if (j == 0) {
             // if the first segment - to account for the range boundary
+#pragma omp parallel for
             for (int i = 0; i < size; i++) {
                 if (likely(I(i) > jI + stepI)) continue;  // wi = 0;
                 if (likely(I(i) > jI)) {
@@ -307,6 +328,7 @@ void fastBilateralFilter(const pfs::Array2Df &I, pfs::Array2Df &J,
             }
         } else if (j == NB_SEGMENTS - 1) {
             // if the last segment - to account for the range boundary
+#pragma omp parallel for
             for (int i = 0; i < size; i++) {
                 if (likely(I(i) < jI - stepI)) continue;  // wi = 0;
                 if (likely(I(i) < jI)) {
@@ -316,6 +338,7 @@ void fastBilateralFilter(const pfs::Array2Df &I, pfs::Array2Df &J,
                     J(i) += (*JJ)(i);
             }
         } else {
+#pragma omp parallel for
             for (int i = 0; i < size; i++) {
                 float wi = (stepI - fabs(I(i) - jI)) / stepI;
                 if (unlikely(wi > 0.0f)) J(i) += (*JJ)(i)*wi;

```

**Comment by fcomida** [comment]
It's durand.

**Comment by fcomida** [comment]
I vectorized other parts, your changes are complementary to mine.

**Comment by heckflosse** [comment]
great :)

**Comment by heckflosse** [comment]
writing vecorized code (even readable vectorized code) can be very easy ;-P

**Comment by heckflosse** [comment]
@fcomida Franco, here's an enhanced speedup for `fastbilateral.cpp`

It solves some strange memory access patterns and so makes better use of cache

```
diff --git a/src/TonemappingOperators/durand02/fastbilateral.cpp b/src/TonemappingOperators/durand02/fastbilateral.cpp
index 726833c4..9742ba66 100644
--- a/src/TonemappingOperators/durand02/fastbilateral.cpp
+++ b/src/TonemappingOperators/durand02/fastbilateral.cpp
@@ -43,6 +43,10 @@
 #define likely(x) (x)
 #define unlikely(x) (x)
 #endif
+#define BENCHMARK
+#include "../../StopWatch.h"
+#include "../../sleef.c"
+#include "../../opthelper.h"
 
 using namespace std;
 
@@ -59,55 +63,66 @@ class GaussianBlur {
    public:
     GaussianBlur(int nx, int ny, float sigma) : sigma(sigma) {
         init_fftw();
-        int ox = nx;
-        int oy = ny / 2 + 1;  // saves half of the data
+        int oy = ny;
+        int ox = nx / 2 + 1;  // saves half of the data
         const int osize = ox * oy;
         FFTW_MUTEX::fftw_mutex_plan.lock();
-        source = (float *)fftwf_malloc(sizeof(float) * nx * 2 * (ny / 2 + 1));
+        source = (float *)fftwf_malloc(sizeof(float) * ny * 2 * (nx / 2 + 1));
         freq = (fftwf_complex *)fftwf_malloc(sizeof(fftwf_complex) * osize);
         if (source == NULL || freq == NULL) {
             std::bad_alloc excep;
             FFTW_MUTEX::fftw_mutex_plan.unlock();
             throw excep;
         }
-        fplan_fw = fftwf_plan_dft_r2c_2d(nx, ny, source, freq, FFTW_ESTIMATE);
-        fplan_in = fftwf_plan_dft_c2r_2d(nx, ny, freq, source, FFTW_ESTIMATE);
+        fplan_fw = fftwf_plan_dft_r2c_2d(ny, nx, source, freq, FFTW_ESTIMATE);
+        fplan_in = fftwf_plan_dft_c2r_2d(ny, nx, freq, source, FFTW_ESTIMATE);
         FFTW_MUTEX::fftw_mutex_plan.unlock();
     }
 
     void blur(const pfs::Array2Df &I, pfs::Array2Df &J) {
+        BENCHFUN
         int x, y;
 
         int nx = I.getCols();
         int ny = I.getRows();
         int nsize = nx * ny;
 
-        int ox = nx;
-        int oy = ny / 2 + 1;  // saves half of the data
+        int oy = ny;
+        int ox = nx / 2 + 1;  // saves half of the data
 
+        #pragma omp parallel for
         for (y = 0; y < ny; y++)
-            for (x = 0; x < nx; x++) source[x * ny + y] = I(x, y);
+            for (x = 0; x < nx; x++)
+                source[y * nx + x] = I(x, y);
 
         fftwf_execute(fplan_fw);
 
         // filter
         float sig = nx / (2.0f * sigma);
         float sig2 = 2.0f * sig * sig;
-        for (x = 0; x < ox / 2; x++)
-            for (y = 0; y < oy; y++) {
-                float d2 = x * x + y * y;
-                float kernel = exp(-d2 / sig2);
-
-                freq[x * oy + y][0] *= kernel;
-                freq[x * oy + y][1] *= kernel;
-                freq[(ox - x - 1) * oy + y][0] *= kernel;
-                freq[(ox - x - 1) * oy + y][1] *= kernel;
+        StopWatch Stope("exp loop");
+        #pragma omp parallel for
+        for (y = 0; y < oy / 2; y++) {
+            float yy = y * y;
+            for (x = 0; x < ox; x++) {
+                float d2 = x * x + yy;
+                float kernel = xexpf(-d2 / sig2);
+
+                freq[y * ox + x][0] *= kernel;
+                freq[y * ox + x][1] *= kernel;
+                freq[(oy - y - 1) * ox + x][0] *= kernel;
+                freq[(oy - y - 1) * ox + x][1] *= kernel;
             }
-
+        }
+        Stope.stop();
         fftwf_execute(fplan_in);
 
-        for (x = 0; x < nx; x++)
-            for (y = 0; y < ny; y++) J(x, y) = source[x * ny + y] / nsize;
+        StopWatch Stopl("last loop");
+        #pragma omp parallel for
+        for (y = 0; y < ny; y++)
+            for (x = 0; x < nx; x++) 
+            J(x, y) = source[y * nx + x] / nsize;
+        Stopl.stop();
     }
 
     ~GaussianBlur() {
@@ -243,17 +258,9 @@ void fastBilateralFilter(const pfs::Array2Df &I, pfs::Array2Df &J,
     }
 
     pfs::Array2Df *JJ;
-    //  if( downsample != 1 )
-    //    JJ = new pfs::Array2D(w,h);
-
-    //  w /= downsample;
-    //  h /= downsample;
 
     int sizeZ = w * h;
-    //  pfs::Array2D* Iz = new pfs::Array2D(w,h);
-    //  downsampleArray(I,Iz);
     const pfs::Array2Df *Iz = &I;
-    //  sigma_s /= downsample;
 
     pfs::Array2Df jJ(w, h);
     pfs::Array2Df jG(w, h);
@@ -272,11 +279,29 @@ void fastBilateralFilter(const pfs::Array2Df &I, pfs::Array2Df &J,
 
         float jI = minI + j * stepI;  // current intensity value
 
-        for (int i = 0; i < sizeZ; i++) {
-            float dI = (*Iz)(i)-jI;
-            jG(i) = exp(-(dI * dI) / (sigma_r * sigma_r));
-            jH(i) = jG(i) * I(i);
+#pragma omp parallel
+{
+#ifdef __SSE2__
+        vfloat sqrsigma_rv = F2V(sigma_r * sigma_r);
+        vfloat jIv = F2V(jI);
+#endif
+        #pragma omp for
+        for (int i = 0; i < h; i++) {
+            int j = 0;
+#ifdef __SSE2__
+            for (; j < w-3; j+=4) {
+                vfloat dIv = LVFU((*Iz)(j, i)) - jIv;
+                STVFU(jG(j, i), xexpf(-(dIv * dIv) / sqrsigma_rv));
+                STVFU(jH(j, i), LVFU(jG(j, i)) * LVFU(I(j, i)));
+            }
+#endif
+            for (; j < w; j++) {
+                float dI = (*Iz)(j, i) - jI;
+                jG(j, i) = xexpf(-(dI * dI) / (sigma_r * sigma_r));
+                jH(j, i) = jG(j, i) * I(j, i);
+            }
         }
+}
 
         gaussian_blur.blur(jG, jK);
         gaussian_blur.blur(jH, jH);
@@ -284,12 +309,13 @@ void fastBilateralFilter(const pfs::Array2Df &I, pfs::Array2Df &J,
         //    convolveArray(jG, sigma_s, jK);
         //    convolveArray(jH, sigma_s, jH);
 
-        for (int i = 0; i < sizeZ; i++)
-            if (likely(jK(i) != 0.0f))
-                jJ(i) = jH(i) / jK(i);
-            else
-                jJ(i) = 0.0f;
-
+        #pragma omp parallel for
+        for (int i = 0; i < h; i++) {
+            for (int j = 0; j < w; j++) {
+                float temp = jK(j, i);
+                jJ(j, i) = temp != 0.f ? jH(j, i) / temp : 0.f;
+            }
+        }
         //  if( downsample == 1 )
         JJ = &jJ;  // No upsampling is necessary
                    //    else
@@ -297,6 +323,7 @@ void fastBilateralFilter(const pfs::Array2Df &I, pfs::Array2Df &J,
 
         if (j == 0) {
             // if the first segment - to account for the range boundary
+#pragma omp parallel for
             for (int i = 0; i < size; i++) {
                 if (likely(I(i) > jI + stepI)) continue;  // wi = 0;
                 if (likely(I(i) > jI)) {
@@ -307,6 +334,7 @@ void fastBilateralFilter(const pfs::Array2Df &I, pfs::Array2Df &J,
             }
         } else if (j == NB_SEGMENTS - 1) {
             // if the last segment - to account for the range boundary
+#pragma omp parallel for
             for (int i = 0; i < size; i++) {
                 if (likely(I(i) < jI - stepI)) continue;  // wi = 0;
                 if (likely(I(i) < jI)) {
@@ -316,6 +344,7 @@ void fastBilateralFilter(const pfs::Array2Df &I, pfs::Array2Df &J,
                     J(i) += (*JJ)(i);
             }
         } else {
+#pragma omp parallel for
             for (int i = 0; i < size; i++) {
                 float wi = (stepI - fabs(I(i) - jI)) / stepI;
                 if (unlikely(wi > 0.0f)) J(i) += (*JJ)(i)*wi;

```

But we also should try to use a faster GaussianBlur. The current one needs about 0.4 seconds for a 24 Mpixel file while the implementation in RT needs just 0.06 seconds for the same input using the same sigma.

Ingo

**Comment by fcomida** [comment]
Ingo it seems there's an issue with "portrait" images in previews with
drago and durand. A whole stripe on the right is either black or white
(only in previews) and calculating those previews is very slow. I'll update
you later...

**Comment by heckflosse** [comment]
Franco, I can not confirm the issue with `portrait` images on drago_speedup branch. `My` local durand changes also work fine with portrait images.

**Comment by heckflosse** [comment]
Franco, for information. I have durand down to 1.3 seconds for a 24 MPixel file (is 51.5 seconds with current dev code)

**Comment by fcomida** [comment]
Ingo that's impressive. There's definitely something wrong with my
environment, especially with previews, especially with "portrait" images in
previews



sometimes I got a completely white preview with reinhard02,
randomly......aaaaargh :D ​

**Comment by heckflosse** [comment]
@fcomida Franco, I created a [new branch](https://github.com/heckflosse/LuminanceHDR/tree/durand_speedup) with the durand optimizations in my fork

**Comment by fcomida** [comment]
Thanks, good job.
Almost exactly what I did myself, apart from the trick for gaussianblur.

**Comment by heckflosse** [comment]
@fcomida Franco, because some of the tonemappers use `findMaxMinPercentile` and each one has it's own implementation, I suggest to make an own compilation unit for that and also make it working correctly independant on the range of input values (as I already did by the scaling in my durand branch). I'm currently doing this for the `findMaxMinPercentile` we use in RawTherapee for fattal. I'm also removing the LUT.h dependency.
I will report when that's completed.

Ingo

**Comment by fcomida** [comment]
@heckflosse That's a good idea. I cloned you branch and it's running fine,
I'm gonna diffing both trees (yours and mine) to see where i screwed up.
Thanks again for your help.

**Comment by heckflosse** [comment]
@fcomida Franco, [about findMaxMinPercentile](https://github.com/Beep6581/RawTherapee/issues/4195)

**Comment by fcomida** [comment]
@heckflosse Yes I had already adapted your implementation in fattal for
durand, as you did as well. And gaussian blur is used in fattal too.

**Comment by heckflosse** [comment]
@fcomida Oh, very good 👍

**Comment by heckflosse** [comment]
@fcomida Franco, about Mantiuk06: The key to speedup this tonemapper imho is

1) use the sleef functions for exp, log and pow
2) get rid of the std::transform stuff to make faster omp parallelized versions
3) then also use the vectorized xexpf and xlogf from sleefsseavx.*

Just a first thought...

**Comment by fcomida** [comment]
@heckflosse if we (you) can speed up mantiuk06 just like durand it would be
amazing. Togheter with fattal it's the most loved operator by large...I
like ferradans a lot too.

**Comment by fcomida** [comment]
@heckflosse Ingo, initial speed up of mantiuk06 from 27sec to 19sec for a
14MP image. I fixed previews :O :/ :(

**Comment by heckflosse** [comment]
@fcomida Franco, here's a small speedup for `denormalizeRGB` in `mantiuk06\contrast_domain.cpp`

```C
template <typename T>
inline T decodefast(const T &value) {
    if (value <= -5.766466716f) {
        return xexpf(value) * 12.92f;
    }
    return 1.055f * xexpf(value / 2.4f) - 0.055f;
}

void denormalizeRGB(Array2Df &R, Array2Df &G, Array2Df &B, const Array2Df &Y,
                    float saturationFactor) {
    const int size = static_cast<int>(Y.size());

/* Transform to sRGB */
#pragma omp parallel for
    for (int j = 0; j < size; j++) {
        float myY = Y(j) * 2.30258509f;
        R(j) = decodefast(xlogf(R(j)) * saturationFactor + myY);
        G(j) = decodefast(xlogf(G(j)) * saturationFactor + myY);
        B(j) = decodefast(xlogf(B(j)) * saturationFactor + myY);
    }
}
```

**Comment by heckflosse** [comment]
@fcomida Franco, about ferradans. It shoots a large armada of fftwf executes. In the fattal implementation of rt @agriggio found a way to [speed up the fftwf executes](https://github.com/Beep6581/RawTherapee/blob/dev/rtengine/tmo_fattal02.cc#L1221). Maybe you can this method also for ferrandans.

**Comment by fcomida** [comment]
@heckflosse I'll have a look at it. Regarding mantiuk06,
PyramidT::multipy() and computeAndAddDivergence(...) can be vectorized but
the underline data structure (XYGradient) must be changed, my next task.

**Comment by Benitoite** [comment]
https://filebin.net/j0kut21qp11nit0g/Luminance_HDR_2.5.2-SNAPSHOT-durandTest1.dmg
@fcomida  @clanmills Mac build of @heckflosse's Durand speedup seems to work fast.

PS: Discuss user has posted a pretty amazing example of the LHDR Ferradans '11 method:
https://discuss.pixls.us/t/salvaging-a-raw/5769

**Comment by heckflosse** [comment]
@fcomida Franco,

> PyramidT::multipy() and computeAndAddDivergence(...) can be vectorized but the underline data structure (XYGradient) must be changed

Maybe the L2CVFU macro will be usefull there to avoid changing the underlying data structure.

**Comment by fcomida** [comment]
I like the way ferradans treats colors. In many pictures, even the original
ones, not processed by lhdr, you will say there's some color cast in them
after a comparison with the one processed by ferradans. It separates colors
well, I don't have a technical way for explaining this behaviour.

**Comment by heckflosse** [comment]
@fcomida ferradans is a beast concerning opimizations :( Still trying.
I like it too!

**Comment by heckflosse** [comment]
Fixed with 0b394f7
